File: apps/translate-tencent_HY-MT/Android/app/src/main/java/ai/zetic/demo/offlinetranslator/model/Language.kt
package ai.zetic.demo.offlinetranslator.model

import java.util.Locale

/**
 * A translation language. `auto` is a source-only sentinel meaning "let the model infer the
 * source language" — the prompt only ever needs the *target*, so Detect needs no special handling.
 * Mirrors `Models/Language.swift`.
 */
data class Language(
    val id: String,          // stable identifier / rough BCP-47 code
    val englishName: String, // name used when building the model prompt
    val nativeName: String,  // endonym shown in the picker
) {
    val isDetect: Boolean get() = id == "auto"

    /** BCP-47-ish locale for Android TextToSpeech voice selection. */
    val speechLocale: Locale
        get() = when (id) {
            "auto", "en" -> Locale.US
            "zh-Hans" -> Locale.SIMPLIFIED_CHINESE
            "zh-Hant" -> Locale.TRADITIONAL_CHINESE
            "nb" -> Locale("nb", "NO")
            else -> Locale.forLanguageTag(id)
        }

    companion object {
        /** Source-only "Detect language" sentinel. */
        val detect = Language("auto", "Detect language", "Detect language")

        /**
         * The languages the model handles well; ordered alphabetically by English name.
         * Verbatim from the iOS `Language.all`.
         */
        val all: List<Language> = listOf(
            Language("ar", "Arabic", "العربية"),
            Language("bn", "Bengali", "বাংলা"),
            Language("bg", "Bulgarian", "Български"),
            Language("zh-Hans", "Chinese (Simplified)", "简体中文"),
            Language("zh-Hant", "Chinese (Traditional)", "繁體中文"),
            Language("cs", "Czech", "Čeština"),
            Language("da", "Danish", "Dansk"),
            Language("nl", "Dutch", "Nederlands"),
            Language("en", "English (US)", "English (US)"),
            Language("fi", "Finnish", "Suomi"),
            Language("fr", "French", "Français"),
            Language("de", "German", "Deutsch"),
            Language("el", "Greek", "Ελληνικά"),
            Language("he", "Hebrew", "עברית"),
            Language("hi", "Hindi", "हिन्दी"),
            Language("hu", "Hungarian", "Magyar"),
            Language("id", "Indonesian", "Indonesia"),
            Language("it", "Italian", "Italiano"),
            Language("ja", "Japanese", "日本語"),
            Language("ko", "Korean", "한국어"),
            Language("ms", "Malay", "Melayu"),
            Language("nb", "Norwegian", "Norsk"),
            Language("fa", "Persian", "فارسی"),
            Language("pl", "Polish", "Polski"),
            Language("pt", "Portuguese", "Português"),
            Language("ro", "Romanian", "Română"),
            Language("ru", "Russian", "Русский"),
            Language("sk", "Slovak", "Slovenčina"),
            Language("es", "Spanish", "Español"),
            Language("sv", "Swedish", "Svenska"),
            Language("ta", "Tamil", "தமிழ்"),
            Language("te", "Telugu", "తెలుగు"),
            Language("th", "Thai", "ไทย"),
            Language("tr", "Turkish", "Türkçe"),
            Language("uk", "Ukrainian", "Українська"),
            Language("vi", "Vietnamese", "Tiếng Việt"),
        )

        /** Source options: "Detect language" first, then every concrete language. */
        val sourceOptions: List<Language> = listOf(detect) + all

        /** Target options: concrete languages only (no "Detect language"). */
        val targetOptions: List<Language> = all

        fun named(id: String): Language = all.firstOrNull { it.id == id } ?: detect
    }
}

File: apps/translate-tencent_HY-MT/iOS/OfflineTranslator/Models/Language.swift
import Foundation

/// A translation language. `auto` is a source-only sentinel meaning "let the model
/// infer the source language" — Hunyuan-MT does this naturally because its prompt
/// only needs the *target* language.
struct Language: Identifiable, Hashable {
    let id: String          // stable identifier / rough BCP-47 code
    let englishName: String // name used when building the model prompt
    let nativeName: String  // endonym shown in the picker

    var isDetect: Bool { id == "auto" }

    /// Source-only "Detect language" sentinel.
    static let detect = Language(id: "auto", englishName: "Detect language", nativeName: "Detect language")
}

extension Language {
    /// The languages Tencent HY-MT handles well. Kept broad on purpose so the
    /// live demo can show off coverage; ordered alphabetically by English name.
    static let all: [Language] = [
        Language(id: "ar", englishName: "Arabic", nativeName: "العربية"),
        Language(id: "bn", englishName: "Bengali", nativeName: "বাংলা"),
        Language(id: "bg", englishName: "Bulgarian", nativeName: "Български"),
        Language(id: "zh-Hans", englishName: "Chinese (Simplified)", nativeName: "简体中文"),
        Language(id: "zh-Hant", englishName: "Chinese (Traditional)", nativeName: "繁體中文"),
        Language(id: "cs", englishName: "Czech", nativeName: "Čeština"),
        Language(id: "da", englishName: "Danish", nativeName: "Dansk"),
        Language(id: "nl", englishName: "Dutch", nativeName: "Nederlands"),
        Language(id: "en", englishName: "English (US)", nativeName: "English (US)"),
        Language(id: "fi", englishName: "Finnish", nativeName: "Suomi"),
        Language(id: "fr", englishName: "French", nativeName: "Français"),
        Language(id: "de", englishName: "German", nativeName: "Deutsch"),
        Language(id: "el", englishName: "Greek", nativeName: "Ελληνικά"),
        Language(id: "he", englishName: "Hebrew", nativeName: "עברית"),
        Language(id: "hi", englishName: "Hindi", nativeName: "हिन्दी"),
        Language(id: "hu", englishName: "Hungarian", nativeName: "Magyar"),
        Language(id: "id", englishName: "Indonesian", nativeName: "Indonesia"),
        Language(id: "it", englishName: "Italian", nativeName: "Italiano"),
        Language(id: "ja", englishName: "Japanese", nativeName: "日本語"),
        Language(id: "ko", englishName: "Korean", nativeName: "한국어"),
        Language(id: "ms", englishName: "Malay", nativeName: "Melayu"),
        Language(id: "nb", englishName: "Norwegian", nativeName: "Norsk"),
        Language(id: "fa", englishName: "Persian", nativeName: "فارسی"),
        Language(id: "pl", englishName: "Polish", nativeName: "Polski"),
        Language(id: "pt", englishName: "Portuguese", nativeName: "Português"),
        Language(id: "ro", englishName: "Romanian", nativeName: "Română"),
        Language(id: "ru", englishName: "Russian", nativeName: "Русский"),
        Language(id: "sk", englishName: "Slovak", nativeName: "Slovenčina"),
        Language(id: "es", englishName: "Spanish", nativeName: "Español"),
        Language(id: "sv", englishName: "Swedish", nativeName: "Svenska"),
        Language(id: "ta", englishName: "Tamil", nativeName: "தமிழ்"),
        Language(id: "te", englishName: "Telugu", nativeName: "తెలుగు"),
        Language(id: "th", englishName: "Thai", nativeName: "ไทย"),
        Language(id: "tr", englishName: "Turkish", nativeName: "Türkçe"),
        Language(id: "uk", englishName: "Ukrainian", nativeName: "Українська"),
        Language(id: "vi", englishName: "Vietnamese", nativeName: "Tiếng Việt"),
    ]

    /// Source options: "Detect language" first, then every concrete language.
    static let sourceOptions: [Language] = [.detect] + all

    /// Target options: concrete languages only (no "Detect language").
    static let targetOptions: [Language] = all

    static func named(_ id: String) -> Language {
        all.first { $0.id == id } ?? detect
    }

    /// BCP-47-ish code for AVSpeechSynthesizer voice selection.
    var speechCode: String {
        switch id {
        case "auto": return "en-US"
        case "en": return "en-US"
        case "zh-Hans": return "zh-CN"
        case "zh-Hant": return "zh-TW"
        case "nb": return "nb-NO"
        default: return id
        }
    }
}
